packages/vitest-pool-workers/src/pool/helpers.ts

```diff
@@ -24,3 +25,56 @@ export function getRelativeProjectPath(
 		return path.relative("", projectPath);
 	}
 }
+
+export function calculateAvailableThreads() {
+	return os.availableParallelism ? os.availableParallelism() : os.cpus().length;
+}
+
+type Resolve = () => void;
+
+/**
+ * A generic pool that can be used to limit the number of active resources. The resources
+ * are anonymous, so we don't care about their implementation. The goal is just to count
+ * the number of resources in-use.
+
+ */
+export class ResourcePool {
```

Comment:
I wonder if this should actually be called a `Semaphore` (see https://learn.microsoft.com/en-us/dotnet/api/system.threading.semaphore?view=net-8.0).

Reply:
That looks pretty accurate.

https://en.wikipedia.org/wiki/Semaphore_(programming)

> A useful way to think of a semaphore as used in a real-world system is as a record of how many units of a particular resource are available, coupled with operations to adjust that record safely (i.e., to avoid [race conditions](https://en.wikipedia.org/wiki/Race_condition)) as units are acquired or become free, and, if necessary, wait until a unit of the resource becomes available.